metrics/metrics.go

```diff
@@ -4,21 +4,74 @@ import (
 	"fmt"
 	"net/http"
 	_ "net/http/pprof" // adds default pprof endpoint at /debug/pprof
-	"strconv"
+	"runtime"
 
+	"github.com/drand/drand/log"
 	"github.com/prometheus/client_golang/prometheus"
-	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
 var (
-	APICallCounter = promauto.NewCounterVec(prometheus.CounterOpts{
+	// PrivateMetrics about the internal world (go process, private stuff)
+	PrivateMetrics = prometheus.NewRegistry()
+	// HTTPMetrics about the public surface area (http requests, cdn stuff)
+	HTTPMetrics = prometheus.NewRegistry()
+	// GroupMetrics about the group surface (grp, group-member stuff)
+	GroupMetrics = prometheus.NewRegistry()
+
+	// APICallCounter (Group) how many grpc calls
+	APICallCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
 		Name: "api_call_counter",
 		Help: "Number of API calls that we have received",
 	}, []string{"api_method"})
+	// GroupDialFailures (Group) how manuy failures connecting outbound
+	GroupDialFailures = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "dial_failures",
+		Help: "Number of times there have been network connection issues",
+	}, []string{"peer_address"})
+	// GroupConnections (Group) how many GrpcClient connections are present
+	GroupConnections = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "group_connections",
+		Help: "Number of peers with current GrpcClient connections",
+	})
+
+	metricsBound = false
 )
 
+func bindMetrics() {
+	if metricsBound {
+		return
+	}
+	metricsBound = true
+
+	// The private go-level metrics live in private.
+	PrivateMetrics.Register(prometheus.NewGoCollector())
+	PrivateMetrics.Register(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
+
+	// Group metrics
+	group := []prometheus.Collector{
+		APICallCounter,
+		GroupDialFailures,
+		GroupConnections,
+	}
+	for _, c := range group {
+		GroupMetrics.Register(c)
+		PrivateMetrics.Register(c)
+	}
+}
+
 // Start starts a prometheus metrics server with debug endpoints.
-func Start(metricsPort int) {
-	fmt.Printf("drand: starting metrics server on port %v", metricsPort)
-	fmt.Printf("%v", http.ListenAndServe(":"+strconv.Itoa(metricsPort), nil))
+func Start(metricsBind string) {
+	log.DefaultLogger.Debug("metrics", "private listener started", "at", metricsBind)
+	bindMetrics()
+
+	s := http.Server{Addr: metricsBind}
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.HandlerFor(PrivateMetrics, promhttp.HandlerOpts{Registry: PrivateMetrics}))
+	mux.HandleFunc("/debug/gc", func(w http.ResponseWriter, req *http.Request) {
+		runtime.GC()
```

Comment:
Can you write down a part in the deployment section about the whole promotheus feature, and specifically list the endpoints (or groups) that are private and the one that are public ? 
Reason I'm asking is to let operators know it's there but also  I wouldn't want operators  expose `runtime.GC()` publicly for example so we must be extra clear on this.

Reply:
note that this currently is the same `--metrics` private endpoint that was added by cloudflare earlier. I don't think we've decided how to expose the 'group-visible' metrics.
My best guess was we'll recommend to keep this `--metrics` port private, and push all metrics to it for local collection.

In addition, we'll take the 'group' metrics, and provide them over the GRPC interface somehow. haven't figured that out yet :)

Reply:
All right sounds good !
Only thing then is to just correct the [doc](https://github.com/drand/drand/blob/master/docs/DEPLOYMENT.md#metrics) to say it can be used with a `address:port` style.